File: lib/std/network.lua
require 'std.core.util'
require 'std.core.types.player'
require 'std.core.game'
require 'std.core.event'

local base64 = require 'std.ext.base64'

network = network or {}

-- max amount of bytes allowed per data chunk
local MAX_PER_CHUNK = 255 - 6
-- max amount of chunks per send "round"
local MAX_CHUNKS_PER_ROUND = 4

-- max amount of concurrent transmissions each player can have
local MAX_TRANSMISSIONS_PER_PLAYER = 1
-- max amount of total concurrent transmissions allowed in the game
local MAX_TRANSMISSIONS_PER_GAME = 4

local STATUS_QUEUED_START = 0
local STATUS_QUEUED_CHUNK = 1
local STATUS_INFLIGHT = 2

local networkContext = util.context("network", {
    listeners = {},
    transmissions = {},
    queued = {},
    counters = {}
})

local transmissions = networkContext.transmissions
local listeners = networkContext.listeners
local counters = networkContext.counters
local queuedTransmissions = networkContext.queued

function network.receive(id, callback)
    ASSERT_ARG_TYPE(1, "id", "string")
    ASSERT_ARG_TYPE(2, "callback", "function")

    if listeners[id] then
        listeners[id] = callback
    else
        local trig = CreateTrigger()
        for _, v in pairs(player.getAll()) do
            BlzTriggerRegisterPlayerSyncEvent(trig, v.__h, id, false)
        end
        TriggerAddAction(trig, function()
            local p = __WrapPlayer(GetTriggerPlayer())
            local d = BlzGetTriggerSyncData()
            ceres.catch(listeners[id], p, d)
        end)
        listeners[id] = callback
    end
end

function network.send(id, payload)
    ASSERT_ARG_TYPE(1, "id", "string")
    ASSERT_ARG_TYPE(2, "payload", "string")
    ASSERT_COND("#payload <= 255")

    BlzSendSyncData(id, payload)
end

local updateQueue, enqueueTransmission

local function countInFlightTransmissionsOf(sender)
    if not transmissions[sender.id] then
        return 0
    else
        local count = 0
        for _, v in pairs(transmissions[sender.id]) do
            if v.status == STATUS_INFLIGHT then
                count = count + 1
            end
        end
        return count
    end
end

local function countInFlightTransmissionsAll()
    local count = 0
    for _, p in pairs(player.getAll()) do
        count = count + countInFlightTransmissionsOf(p)
    end
    return count
end

local CHUNK_HEADER_PACK=">I3"
local CHUNK_HEADER_SIZE = 4
local function sendTransmissionRound(transmission)
    if transmission.sender.isLocal then
        if game.isSingleplayer() then
            MAX_CHUNKS_PER_ROUND = 16384
        end

        for i=1, MAX_CHUNKS_PER_ROUND do
            local chunkData = table.remove(transmission.queuedChunks, 1)

            if not chunkData then
                break
            end

            local chunkHeader = base64.encode(string.pack(CHUNK_HEADER_PACK, transmission.id))

            network.send("__tc", chunkHeader .. chunkData)
        end
    end

    transmission.status = STATUS_INFLIGHT
end

local function receiveTransmissionChunk(sender, data)
    local transmissionId = string.unpack(CHUNK_HEADER_PACK, base64.decode(data:sub(1, CHUNK_HEADER_SIZE)))
    local chunkData = data:sub(CHUNK_HEADER_SIZE + 1)

    local transmission = transmissions[sender.id][transmissionId]
    table.insert(transmission.received, chunkData)

    if #transmission.received == transmission.queuedCount then
        transmissions[sender.id][transmission.id] = nil

        ceres.catch(transmission.callback, table.concat(transmission.received), "success")
        updateQueue()
        return
    end

    transmission.chunkCounter = transmission.chunkCounter + 1
    if transmission.chunkCounter % MAX_CHUNKS_PER_ROUND == 0 then
        enqueueTransmission(transmission, STATUS_QUEUED_CHUNK)
        updateQueue()
    end
end

local TRANSMISSION_HEADER_PACK = ">I3I3"
local function sendTransmissionHeader(transmission)
    if transmission.sender.isLocal then
        local headerData = base64.encode(string.pack(TRANSMISSION_HEADER_PACK, transmission.id, #transmission.queuedChunks))
        network.send("__th", headerData)
    end

    transmission.status = STATUS_INFLIGHT
end

local function receiveTransmissionHeader(sender, data)
    local transmissionId, chunkCount = string.unpack(TRANSMISSION_HEADER_PACK, base64.decode(data))
    local transmission = transmissions[sender.id][transmissionId]
    transmission.queuedCount = chunkCount

    sendTransmissionRound(transmission)
end

enqueueTransmission = function(transmission, status)
    transmission.status = status
    table.insert(queuedTransmissions, transmission)
end

updateQueue = function()
    if countInFlightTransmissionsAll() >= MAX_TRANSMISSIONS_PER_GAME then
        return
    end

    for k, v in ipairs(queuedTransmissions) do
        if countInFlightTransmissionsOf(v.sender) < MAX_TRANSMISSIONS_PER_PLAYER then
            if v.status == STATUS_QUEUED_START then
                sendTransmissionHeader(v)
            elseif v.status == STATUS_QUEUED_CHUNK then
                sendTransmissionRound(v)
            end
            table.remove(queuedTransmissions, k)

            -- tail call
            return updateQueue()
        end
    end
end

function network.synchronize(sender, payload, callback)
    ASSERT_ARG_TYPE(1, "sender", "player")
    ASSERT_ARG_TYPE(2, "payload", "string")
    ASSERT_ARG_TYPE(3, "callback", "function")

    transmissions[sender.id] = transmissions[sender.id] or {}
    counters[sender.id] = counters[sender.id] or 0
    local senderTransmissions = transmissions[sender.id]
    local transmissionId = counters[sender.id]
    counters[sender.id] = counters[sender.id] + 1

    local transmission = {
        id = transmissionId,
        chunkCounter = 0,
        sender = sender,
        callback = callback,
        received = {}
    }

    senderTransmissions[transmissionId] = transmission

    if sender.isLocal then
        local queued = {}
        for i=0, (#payload // MAX_PER_CHUNK) do
            local segStart = (i * MAX_PER_CHUNK) + 1
            local segEnd = math.min(((i + 1) * MAX_PER_CHUNK), #payload)
            table.insert(queued, payload:sub(segStart, segEnd))
        end

        transmission.queuedChunks = queued
    end

    enqueueTransmission(transmission, STATUS_QUEUED_START)
    updateQueue()
end

network.receive("__th", receiveTransmissionHeader)
network.receive("__tc", receiveTransmissionChunk)

File: ts/src/tsstd/network.ts
import Player from "./core/types/player"
import { isSinglePlayer } from "./core/game"
import * as base64 from "std.ext.base64"

const MAX_PER_CHUNK = 255 - 6
let MAX_CHUNKS_PER_ROUND = 4
const MAX_TRANSMISSIONS_PER_PLAYER = 1
const MAX_TRANSMISSIONS_PER_GAME = 4

enum TransmissionStatus {
    None,
    QueuedStart,
    QueuedChunk,
    Inflight
}

type TransmissionCallback = (payload: string, status: "success" | "failure") => void
type NetworkCallback = (sender: Player, payload: string) => void

interface Transmission {
    id: number
    sender: Player
    received: string[]
    status: TransmissionStatus
    callback: TransmissionCallback
    chunkCounter: number

    queuedCount?: number
    queuedChunks?: string[]
}

let context = util.context("network", {
    listeners: new Map<string, NetworkCallback>(),
    transmissions: [] as Map<number, Transmission>[],
    queued: [] as Transmission[],
    counters: [] as number[]
})

export function receive(id: string, callback: NetworkCallback) {
    if (context.listeners.has(id)) {
        context.listeners.set(id, callback)
    } else {
        let trigger = CreateTrigger()
        for (let player of Player.all) {
            BlzTriggerRegisterPlayerSyncEvent(trigger, player.handle, id, false)
        }
        TriggerAddAction(
            trigger,
            ceres.wrapSafeCall(() => {
                context.listeners.get(id)!(Player.of(GetTriggerPlayer()), BlzGetTriggerSyncData())
            })
        )
        context.listeners.set(id, callback)
    }
}

export function send(id: string, payload: string) {
    assert(payload.length <= 255, "payload must be under 256 chars")
    BlzSendSyncData(id, payload)
}

function countInFlightTransmissions(player: Player): number {
    if (!context.transmissions[player.id]) {
        return 0
    } else {
        let count = 0
        for (let [_, transmission] of context.transmissions[player.id]) {
            if (transmission.status == TransmissionStatus.Inflight) {
                count++
            }
        }
        return count
    }
}

function countInFlightTransmissionsAll(): number {
    let count = 0
    for (let player of Player.all) {
        count += countInFlightTransmissions(player)
    }
    return count
}

const CHUNK_HEADER_PACK = ">I3"
const CHUNK_HEADER_SIZE = 4
function sendTransmissionRound(transmission: Transmission) {
    if (transmission.sender.isLocal) {
        if (isSinglePlayer()) {
            MAX_CHUNKS_PER_ROUND = 16384
        }

        for (let i = 0; i < MAX_CHUNKS_PER_ROUND; i++) {
            let chunkData = table.remove(transmission.queuedChunks!, 1)

            if (!chunkData) {
                break
            }

            let chunkHeader = base64.encode(string.pack(CHUNK_HEADER_PACK, transmission.id))

            send("__tc", chunkHeader + chunkData)
        }
    }

    transmission.status = TransmissionStatus.Inflight
}

function receiveTransmissionChunk(sender: Player, data: string) {
    let [transmissionId] = string.unpack(
        CHUNK_HEADER_PACK,
        base64.decode(data.substring(0, CHUNK_HEADER_SIZE))
    )
    let chunkData = data.substring(CHUNK_HEADER_SIZE)

    let transmission = context.transmissions[sender.id].get(transmissionId)!
    table.insert(transmission.received, chunkData)

    if (transmission.received.length == transmission.queuedCount) {
        context.transmissions[sender.id].delete(transmission.id)

        ceres.safeCall(transmission.callback, table.concat(transmission.received), "success")
        updateQueue()
        return
    }

    transmission.chunkCounter = transmission.chunkCounter + 1
    if (transmission.chunkCounter % MAX_CHUNKS_PER_ROUND == 0) {
        enqueueTransmission(transmission, TransmissionStatus.QueuedChunk)
        updateQueue()
    }
}

const TRANSMISSION_HEADER_PACK = ">I3I3"
function sendTransmissionHeader(transmission: Transmission) {
    if (transmission.sender.isLocal) {
        let headerData = base64.encode(
            string.pack(
                TRANSMISSION_HEADER_PACK,
                transmission.id,
                transmission.queuedChunks?.length
            )
        )
        send("__th", headerData)
    }

    transmission.status = TransmissionStatus.Inflight
}

function receiveTransmissionHeader(sender: Player, data: string) {
    let [transmissionId, chunkCount] = string.unpack(
        TRANSMISSION_HEADER_PACK,
        base64.decode(data)
    ) as [number, number]
    let transmission = context.transmissions[sender.id].get(transmissionId)!
    transmission.queuedCount = chunkCount

    sendTransmissionRound(transmission)
}

function enqueueTransmission(transmission: Transmission, status: TransmissionStatus) {
    transmission.status = status
    table.insert(context.queued, transmission)
}

function updateQueue(): void {
    if (countInFlightTransmissionsAll() >= MAX_TRANSMISSIONS_PER_GAME) {
        return
    }

    for (let [k, v] of ipairs(context.queued)) {
        if (countInFlightTransmissions(v.sender) < MAX_TRANSMISSIONS_PER_PLAYER) {
            if (v.status == TransmissionStatus.QueuedStart) {
                sendTransmissionHeader(v)
            } else if (v.status == TransmissionStatus.QueuedChunk) {
                sendTransmissionRound(v)
            }
            table.remove(context.queued, k)

            // tail call
            return updateQueue()
        }
    }
}

export function synchronize(sender: Player, payload: string, callback: TransmissionCallback) {
    if (!context.transmissions[sender.id]) {
        context.transmissions[sender.id] = new Map()
    }
    context.counters[sender.id] = context.counters[sender.id] || 0
    let senderTransmissions = context.transmissions[sender.id]
    let transmissionId = context.counters[sender.id]
    context.counters[sender.id]++

    let transmission: Transmission = {
        id: transmissionId,
        chunkCounter: 0,
        sender: sender,
        callback: callback,
        received: [],
        status: TransmissionStatus.None
    }

    senderTransmissions.set(transmissionId, transmission)

    if (sender.isLocal) {
        transmission.queuedChunks = payload.partition(MAX_PER_CHUNK)
    }

    enqueueTransmission(transmission, TransmissionStatus.QueuedStart)
    updateQueue()
}

receive("__th", receiveTransmissionHeader)
receive("__tc", receiveTransmissionChunk)
